Refactor to allow containers to be directly used as spaces

zsunberg opened the pull request:
@findmyway Here is my start to a PR related to #8 . I haven't started implementing anything yet, just wrote down my ideas in the README. Let me know what you think.

@mossr, @mykelk, @jamgochiana, feel free to chime in if you have any comments.

findmyway commented on README.md:
```diff
@@ -5,39 +5,85 @@
+Continuous spaces represent sets that have an uncountable number of elements they have a `SpaceStyle` of type `ContinuousSpaceStyle`. CommonRLSpaces does not adopt a rigorous mathematical definition of a continuous set, but, roughly, elements in the interior of a continuous space have other elements arbitrarily close to them.
+Continuous spaces have two additional interface functions:
+- `bounds(space)` returns upper and lower bounds in a tuple. For example, if `space` is a unit circle, `bounds(space)` will return `([-1.0, -1.0], [1.0, 1.0])`. This allows agents to choose policies that appropriately cover the space e.g. a normal distribution with a mean of `mean(bounds(space))` and a standard deviation of half the distance between the bounds.
+- [I have a few other ideas, but we can discuss later]
```
Another more common function is `clamp`

findmyway commented on README.md:
```diff
@@ -48,15 +94,14 @@ julia> rand(s) in s
+FiniteSpaceStyle{(2, 3)}()
 
-julia> size(s)
+julia> elsize(s)
```
we may also have `eltype` implemented

findmyway commented on README.md:
```diff
@@ -5,39 +5,85 @@
+
+### Hybrid spaces
+
+[need to figure this out]
```
For *Hybrid Spaces*, I think we can still use `product(s1, s2)` to create a hybrid space.

To help access each sub space easily, `s1` and `s2` may be named spaces and the hybrid space need to support `getindex`.

zsunberg commented:
Ok, I have now added `Box`. A few things that I would like feedback on:

1. I removed the size parameter from the space styles because this seems like a trait that is orthogonal to continuity, and it only really fully makes sense for array spaces.
2. By default, `Box` uses StaticArrays even when constructed with builtin arrays.

findmyway commented:
Looking at the definition. The `Box` seems to be a specialized case of a `HybridSpace`.

I mean it's more or less the same with the result of `product(intervals...)`, except that it provides two extra information on the element:

1. Its element is an `Array`, which also means we know the `eltype` and `size`
2. The bounds are preallocated

Such information are usually environment specific. Some may want a tuple instead of array. Or others do not care about the *preallocation* feature.

This makes me wonder that, maybe we can just define some traits to recognize some common elemental  spaces. We don't need to care too much about the container type (the result type of `product`). It may be an `Array`, a `Tuple`, a `NamedTuple`, a `Dict` or whatever. To sample an element from the hybrid space, we simply create a **similar** container, except that each subspace is replaced with an inner sampling. So that we can still keep the `name`, `type` or `size` info from the container.

In some extreme cases, we can still add a dedicated implementation like `Box` above if we need the *preallocation* feature.

With the traits defiend above, I think this issue can be addressed  https://github.com/JuliaMath/IntervalSets.jl/issues/66

Note that we may need our customized definition of `rand`, `in`, `eltype` and `elsize` here to avoid type piracy.

```julia
rand(x) = rand(rng, x)
rand(rng, x) = rand(rng, SpaceStyle(x), x)

rand(rng, ::DiscreteSpace, x) = Random.rand(rng, x)
rand(rng, ::ContinuousSpace, x) = ...

rand(rng, ::Any, xs) = map(x -> rand(rng, x), xs)
```

This means `product` and `HygirdSpace` are replaced with more concrete constructors, like `tuple` and `Tuple{DiscreteSpace, ContinuousSpace}`.

albheim commented on src/product.jl:
```diff
@@ -0,0 +1,23 @@
+# handle case of 3 or more
+product(s1, s2, s3, args...) = product(product(s1, s2), s3, args...)
+
+struct GenericrSpaceProduct{T<:Tuple}
```
`GenericrSpaceProduct` -> `GenericSpaceProduct`?

zsunberg replied:
Haha, thanks @albheim . The remarkable thing is that i was consistent with the typo throughout.

zsunberg commented:
@findmyway , thanks for the comments! This was very helpful in my thinking about `product`. I agree that we should not care too much about the type that `product` returns.

Do you agree with this statement that I just added to the README: 

> The Cartesian product of two spaces `a` and `b` can be constructed with `c = product(a, b)`.
> 
> The exact form of the resulting space is unspecified and should be considered an implementation detail. The only guarantees are (1) that there will be one unique element of `c` for every combination of one object from `a` and one object from `b` and (2) that the resulting collection conforms to the interface above.
> 
> The `TupleProductSpace` constructor provides a specialized Cartesian product where each element is a tuple, i.e. `TupleProductSpace(a, b)` has elements of type `Tuple{eltype(a), eltype(b)}`.

That is my guiding principle for implementing `product`. I'll try to finish more of the implementation as soon as I get a chance, but it is not finished yet.

By the way, I was thinking that "hybrid" in `HybridSpaceStyle` should refer to whether the space is continuous or discrete, i.e. in addition to `FiniteSpaceStyle` and `ContinuousSpaceStyle`, we would also have `HybridSpaceStyle`. So not every product of two spaces should be a `HybridSpaceStyle`, and in particular `Box` is always `ContinuousSpaceStyle` (see the docstring I added for `Box` for clarification about what it is).

findmyway commented:
Yes, I agree.

And thanks for the explanation of `HybridSpaceStyle`. Now I have a better understanding of *hybrid* here.

zsunberg commented:
Ok, I think this is almost ready to merge, except for a few docstrings and updating the readme. I should be able to add those tomorrow or so. I am very satisfied with how these basics turned out, so I hope everyone agrees that it is a good starting point.

It is currently missing any interface for hybrid continuous-discrete spaces, but we can add that in another chunk. There are also some other issues that I will file later (the two broken tests have raised some questions).

zsunberg commented:
OK, I think this is ready to merge. @findmyway feel free to merge, or I will do it tomorrow if no one objects. I'll add issues tomorrow outlining the things that aren't done yet.
